### PR TITLE
introduce marshaller

### PR DESCRIPTION
**Pull request by BusyJay**
So serialization and deserialization can be customized.

### CODE REVIEW

File: src/codec.rs [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+
+use error::Result;
+
+pub type DeFn<T> = fn(&[u8]) -> Result<T>;
+pub type SerFn<T> = fn(&T, &mut Vec<u8>);
```
Comment (disksing) [inline src/codec.rs]: _MarshalFn_ & _UnmarshalFn_ seems better.

Comment (disksing) [inline src/codec.rs]: Or _SerializeFn_ and _DeserializeFn_.

**Comment by disksing** [comment]
LGTM.

**Comment by siddontang** [comment]
LGTM

**Pull request merged**